Validation error codes with string like object

johnraz opened the pull request:
This one is here to help compare work done in #3716

xordoquy commented:
Adding this to the 3.3.3 milestone to keep on our radars.

tomchristie commented:
Looking feasible at this point. One way to move forward - take a stab at documenting (even if only in the comments) - how does a developer get access to the status codes in an error. Are they visible only if working with the serializer explicitly, or are they available to the existing error handler without any further changes?

johnraz commented:
Error codes are available everywhere the error strings are available, so basically at every level. I Will add decent documentation asap (dont have access to my computer until tomorrow).

johnraz commented:
So to be more specific:

> Are they visible only if working with the serializer explicitly

No.

> Are they visible only if working with the serializer explicitly, or are they available to the existing error handler without any further changes?

Yes. See here: https://github.com/tomchristie/django-rest-framework/pull/3775/files#diff-3a4f34405474c71b1b3c0a0a3b2fbf28R32

johnraz commented:
@tomchristie : do you want me to go more in depth and come up with an actual markdown doc update or is it enough for now? (I'm open to do it, just asking)

tomchristie commented:
@johnraz - It's okay in it's current state - thank you! Still on the backlog, and still queued up behind other stuff that I'm having to focus on ATM - apologies as ever for the snails pace.

johnraz commented:
@tomchristie, @xordoquy rebased, hoping to bump this up a little ;-)

x-warrior commented:
This is great! Do we need any change on this? If you need any help, please let me know! :D

johnraz commented:
it is awaiting final review and :shipit:

johnraz commented:
ping @tomchristie: I put quite a few efforts on this and would really like to see it going forward as I need the feature badly and I'm sure i'm not the only one :smile: 

Just trying to promote the feature a little bit more here :wink:

tomchristie commented:
I'm still lukewarm, but not going to write it off.

`ValidationErrorMessage` is an extra layer of indirection, in a bit of the codebase that lots of users are going to be scanning a lot of the time. Yes it provides an extra bit of functionality, but it also comes at a cost.

It might be useful to see an example of how you'd use this together with some view code or custom exception handler.

johnraz commented:
There is already an example in the tests [here](https://github.com/johnraz/django-rest-framework/blob/validation-error-codes-string-like-object/tests/test_validation_error.py), I can put together a more detailed example if that would help you get your head around it, just let me know.

johnraz commented:
@tomchristie ?

johnraz commented:
@tomchristie: getting back to this one again, rebased on master and also simplified the commit history a bit. 
I will probably squash everything up in the end anyway but it made my life easier for the rebase.
It would be awesome to push this forward for next release 😄

tomchristie commented:
Gotcha, will make sure to consider it for 3.5 yup!

lwm commented:
@johnraz  Thanks for your great work. I am trying to understand your changes a bit more as I would like to see this arrive soon. This PR brings:

1) We can now do this `ValidationError('foo', code='bar')`
2) No breaking changes to existing `ValidationError` handling code
3) A new `ValidationError.code` API to be used as necessary in an `api_exception_handler`.

Is that correct?

The `__str__` for `ValidationErrorMessage` is off (looks like a string!):
```
(Pdb) exc
ValidationError({'integer': ['This field is required.'], 'char': ['This field is required.']},)
(Pdb) type(exc)
<class 'rest_framework.exceptions.ValidationError'>
(Pdb) error
'This field is required.'
(Pdb) type(error)
<class 'rest_framework.exceptions.ValidationErrorMessage'>
```

Finally, is there documentation on the way for this? I could help if needed.

johnraz commented:
@lwm you got the grasp of it yes. The fact that the ValidationErrorMessage looks like a string is made on purpose to avoid breaking ValidationError (point 2). There is no documentation yet.

johnraz commented:
Mmm one thing I overlooked in what you wrote, the new code API is on ValidationErrorMessage.code, not on ValidationError itself.

tomchristie commented:
Related thread: https://groups.google.com/d/msg/django-rest-framework/fGS5VsYy1Xg/F1J2YTIiAwAJ

johnraz commented:
@tomchristie I've just been told the customer I'm working on this for, just started sponsorship on the project 🎉  Hope this will help you push this forward!

tomchristie commented:
👍

tomchristie commented:
Now superseded by #4550.
Thanks so much for both your hard work on this, and your patience.
